### PR TITLE
[SPARK-29761][SQL] do not output leading 'interval' in CalendarInterval.toString

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

remove the leading "interval" in `CalendarInterval.toString`.

### Why are the changes needed?

Although it's allowed to have "interval" prefix when casting string to int, it's not recommended.

This is also consistent with pgsql:
```
cloud0fan=# select interval '1' day;
 interval 
----------
 1 day
(1 row)
```

### Does this PR introduce any user-facing change?

yes, when display a dataframe with interval type column, the result is different.

### How was this patch tested?

updated tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @MaxGekk @dongjoon-hyun @maropu

**Comment by maropu** [comment]
Ur, I see, this change looks reasonable. Actually, it seems pgSQL does so;
```
postgres=# select cast('1 day' as interval);
 interval 
----------
 1 day
(1 row)
```

**Comment by yaooqinn** [comment]
I have been confusing this weird prefix for a long time，but how about we use sql standard output style？

**Comment by cloud-fan** [comment]
AFAIK pgsql has configs to change the interval output string. We can think about it later.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 1]
```diff
@@ -101,18 +101,19 @@ public String toString() {
...
     }
 
+    sb.setLength(sb.length() - 1);
     return sb.toString();
   }
 
   private void appendUnit(StringBuilder sb, long value, String unit) {
     if (value != 0) {
-      sb.append(' ').append(value).append(' ').append(unit).append('s');
+      sb.append(value).append(' ').append(unit).append('s').append(' ');
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: nit: Less `.append()` calls:
```suggestion
      sb.append(value).append(' ').append(unit).append("s ");
```

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 2]
```diff
@@ -101,18 +101,19 @@ public String toString() {
...
-        sb.append(' ').append(s).append(" seconds");
+        sb.append(s).append(" seconds ");
       }
     } else if (months == 0 && days == 0) {
-      sb.append(" 0 microseconds");
+      sb.append("0 microseconds ");
     }
 
+    sb.setLength(sb.length() - 1);
     return sb.toString();
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: Not important but you could leave the code as is, and just do `trim()` on the result.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java [hunk 3]
```diff
@@ -101,18 +101,19 @@ public String toString() {
...
   }
 
   private void appendUnit(StringBuilder sb, long value, String unit) {
```
Comment (MaxGekk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java]: If we always add "s " to the end of units, maybe just pass units with "s "? I mean `"year "`, `"month "`, `"day "` ...

**Review by MaxGekk — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @cloud-fan . I'm +1 for this, but it seems that we need a migration guide in order to avoid surprises because the output change might affect the old Spark application which expects that prefix. (cc @gatorsmile )

This PR already shows that kind of side effects.
```scala
- df.selectExpr(s"d - $i"),
+ df.selectExpr(s"d - INTERVAL'$i'"),
```

**Comment by cloud-fan** [comment]
> This PR already shows that kind of side effects.

The particular case you pointed out is not a breaking change. It relies on `CalendarInterval.toString` which is a private class.

it does change the result when casting interval to string. Let me add a migration guide in case users rely on that string.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**